base/compiler/tfuncs.jl

```diff
@@ -1665,7 +1665,7 @@ function apply_type_nothrow(𝕃::AbstractLattice, argtypes::Vector{Any}, @nospe
                 end
             else
                 istype || return false
-                if !(T <: u.var.ub)
+                if isa(u.var.ub, TypeVar) || !(T <: u.var.ub)
```

Comment:
This could also do `has_free_typevars(u.var.ub)` to cover some other (even more rare) cases